scarlet/blend.py

```diff
@@ -58,6 +59,12 @@ def fit(self, max_iter=200, e_rel=1e-3, **alg_kwargs):
         X = self.parameters
         n_params = len(X)
 
+        # We have to declare the function that inserts sources into the blend with autograd
+        # This has to be done each time we fit a blend,
+        # since the number of components => the number of arguments, which must be linked to the
+        # autograd primitive function
+        defvjp(_add_models, *([partial(_grad_add_models, index=k) for k in range(n_params)]))
```

Comment:
Shouldn't that be declared in `ComponentTree` so that it's available for trees outside of `Blend`?

Reply:
This is the tricky bit about this whole operation. You have to register a `primitive` with autograd with the number of parameters that will be updated by the function with a unique gradient function for each. So we can't do this until we build the blend and know how many components that it has.

My initial run at this (5135580fd15fd9cf7172f5d50f594c058dd01ade) just updated `Box` with a method that updated the gradient each time a model was added, but that still had an overhead because the `model` was being re-written in each step. This cost us ~10% in runtime and with autograd there is no way to avoid this without really going deep into the code and modifying the VJPNode myself (which I started to look into but after a few hours down that rabbit hole this seemed like an easier solution).

Reply:
I don't question the approach itself, but any `ComponentTree` knows the number of components it holds. Are you worried about cases, where we're inserting a ComponentTree into a ComponentTree?

Reply:
That's a good point, I didn't think about that. You're right, this should be in the `ComponentTree` instead of `Blend`.

Reply:
Wait, I take that back. I do remember why I did it this way and you're right in your last comment. `defvjp` updates a global parameter and I wanted to make sure that if someone uses hierarchical components it doesn't get re-written by the child `ComponentTree`s.

Reply:
Then it's brittle, in that the same process cannot declare different `Blend` instances and have them behave consistently

Reply:
True, that's why I put it in `fit` and not `init`. The only way I can see this breaking is if someone has multiple threads all running scarlet at the same time, in the same environment. This is why you don't use global variables.

When we switch to JAX if we run into the same issue it would be a good idea to write a PR that allows us to have local `VJP`s that could be defined for a single `grad` call.

Reply:
But I've been following autograd and I don't think that they're accepting PRs with new features, only bug fixes, to encourage people to use and develop jax. So even if we wanted to write something now they wouldn't accept it.

Reply:
Yeah, I think we both feel where this is headed...